src/phoenix/server/api/routers/v1/datasets.py

```diff
@@ -0,0 +1,294 @@
+                status_code=HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+    try:
+        _check_keys_exist(column_headers, input_keys, output_keys, metadata_keys)
+    except ValueError as e:
+        return Response(
+            content=str(e),
+            status_code=HTTP_422_UNPROCESSABLE_ENTITY,
+        )
+    return Response(
+        background=BackgroundTask(
+            _add_table,
+            request.state.enqueue_for_transaction,
+            get_rows=get_rows,
+            action=action,
+            name=name,
+            description=description,
+            input_keys=list(input_keys),
+            output_keys=list(output_keys),
+            metadata_keys=list(metadata_keys),
+        )
+    )
+
+
+async def _add_table(
+    enqueue: Callable[[DataModification], Awaitable[None]],
+    get_rows: Callable[[], Iterator[Mapping[str, Any]]],
+    input_keys: Sequence[str],
+    output_keys: Sequence[str],
+    metadata_keys: Sequence[str] = (),
+    name: Optional[str] = None,
+    description: Optional[str] = None,
+    action: DatasetTableAction = DatasetTableAction.CREATE,
+) -> None:
+    await enqueue(
+        partial(
+            add_table,
+            table=await run_in_threadpool(get_rows),
+            action=action,
+            name=name,
+            description=description,
+            input_keys=input_keys,
+            output_keys=output_keys,
+            metadata_keys=metadata_keys,
+        )
+    )
+
+
+Name: TypeAlias = Optional[str]
+Description: TypeAlias = Optional[str]
+InputKeys: TypeAlias = FrozenSet[str]
+OutputKeys: TypeAlias = FrozenSet[str]
+MetadataKeys: TypeAlias = FrozenSet[str]
+DatasetId: TypeAlias = int
+
+
+async def _process_csv(
+    content: bytes,
+    content_encoding: Optional[str],
+) -> Tuple[Callable[[], Iterator[Dict[str, Any]]], FrozenSet[str]]:
+    if content_encoding == "gzip":
+        content = await run_in_threadpool(gzip.decompress, content)
+    elif content_encoding == "deflate":
+        content = await run_in_threadpool(zlib.decompress, content)
+    else:
+        raise ValueError(f"Unknown content encoding: {content_encoding}")
+    reader = await run_in_threadpool(lambda c: csv.DictReader(io.StringIO(c.decode())), content)
+    if reader.fieldnames is None:
+        raise ValueError("Missing CSV column header")
+    (header, freq), *_ = Counter(reader.fieldnames).most_common(1)
+    if freq > 1:
+        raise ValueError(f"Duplicated column header in CSV file: {header}")
+    column_headers = frozenset(reader.fieldnames)
+
+    def get_rows() -> Iterator[Dict[str, Any]]:
+        for row in reader:
+            yield row
+
+    return get_rows, column_headers
+
+
+async def _process_pyarrow(
+    content: bytes,
+) -> Tuple[Callable[[], Iterator[Dict[str, Any]]], FrozenSet[str]]:
+    try:
+        reader = pa.ipc.open_stream(content)
+    except pa.ArrowInvalid:
+        raise ValueError("File is not valid pyarrow")
```

Comment:
```suggestion
    except pa.ArrowInvalid as error:
        raise ValueError("File is not valid pyarrow") from error
```

Probably nice to get more details in case of an exception here.